### PR TITLE
PERF: Add a new factor that just computes beta.

### PR DESCRIPTION
**Pull request by ssanderson**
This is a little over 1100x faster (!) than using
`RollingLinearRegressionOfReturns` on my machine.

Profiling output for a 1-month pipeline using both terms for with a
90-day lookback:

```
Tue Nov 21 02:05:48 2017    pipebench/perf/betas.stats

         57724856 function calls (57689155 primitive calls) in 92.342 seconds

   Ordered by: cumulative time
   List reduced from 1212 to 3 due to restriction <'statistical.py'>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       21    0.612    0.029   95.461    4.546 statistical.py:194(compute)
   172201    0.407    0.000   94.843    0.001 statistical.py:201(regress)
       21    0.048    0.002    0.082    0.004 statistical.py:500(compute)
```

- The `compute` function that takes 95 seconds is `RollingLinearRetressionOfReturns`.
- The `compute` function that takes 0.082 seconds is the new implementation.

### CODE REVIEW

**Comment by ssanderson** [comment]
I think the only significant decision that needs to be made here is what we want to do if there are NaNs in the input.  The current implementation uses `nanmean` everywhere possible, which means it produces values more often than the existing `RollingLinearRegressionOfReturns`, (which internally calls scipy.stats.linregress, which is, as it turns out, not particularly efficient when called 172,000 times).

cc @Jstauth @jmccorriston @ahgnaw @twiecki for thoughts on how we want to handle missing data here.  Note that I'm not trying to port our downstream shrinkage beta (I think we should also do that, but the use-case here is to just provide the simplest and fastest reasonable implementation of beta, in particular with an eye toward use in portfolio optimization.)

File: zipline/pipeline/factors/statistical.py [hunk 1]
```diff
@@ -148,11 +155,6 @@ class RollingLinearRegression(CustomFactor, SingleInputMixin):
         The factor/slice whose columns are the predictor/independent variable
         of each regression with `dependent`. If `independent` is a Factor,
         regressions are computed asset-wise.
-    independent : zipline.pipeline.Term with a numeric dtype
```
Comment (ssanderson) [inline zipline/pipeline/factors/statistical.py]: This is an unrelated change to the rest of the PR.  We had this documented twice.

**Comment by ssanderson** [comment]
Hilarious performance note: in the current implementation more than half of the total time is spent constructing namedtuple types:

```
Function                                called...
                                            ncalls  tottime  cumtime
_stats_mstats_common.py:11(linregress)  ->  172201    3.918   15.788  _distn_infrastructure.py:1720(sf)
                                            172201   44.965   58.895  collections.py:305(namedtuple)
                                            344402    0.445    5.133  fromnumeric.py:2796(mean)
                                            172201    2.672   10.022  function_base.py:2298(cov)
                                            344402    0.209    0.272  numeric.py:414(asarray)
                                            172201    0.028    0.028  {len}
```

**Comment by ahgnaw** [comment]
In `empyrical.beta_aligned()` we remove any date-pairs with `np.nan` in them. Here we might want to add a minimum number of non-nan date pairs required to calculate beta.

**Comment by jmccorriston** [comment]
What is the current behavior of `RollingLinearRegressionOfReturns` with regards to nan values?

File: zipline/pipeline/factors/statistical.py [hunk 2]
```diff
@@ -475,3 +477,43 @@ def __new__(cls,
             regression_length=regression_length,
             mask=mask,
         )
+
+
+class SimpleBeta(CustomFactor, StandardOutputs):
+    """
+    Factor producing the slope of a regression line between each asset's daily
+    returns the daily returns of a single "target" asset.
```
Comment (dmichalowicz) [inline zipline/pipeline/factors/statistical.py]: "...each asset's daily returns _and_ the daily returns of..." (missing "and")

**Comment by ssanderson** [comment]
> What is the current behavior of RollingLinearRegressionOfReturns with regards to nan values?

Any asset with a NaN at any point in the lookback window ends up NaN.

**Comment by ssanderson** [comment]
Pushed a change that makes the behavior here match empyrical exactly.  Handling NaNs imposes around a 2.5x performance penalty (from 0.9 seconds per year to 2.2 on my machine).  That's still a huge win compared to 90+ seconds **per month** with the old implementation.

**Before:**
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    3.544    3.544 engine.py:445(compute_chunk)
        2    0.027    0.013    1.469    0.735 mixins.py:189(_compute)
      253    0.518    0.002    0.917    0.004 statistical.py:512(compute)
```

**After:**
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    4.909    4.909 engine.py:445(compute_chunk)
        2    0.028    0.014    2.844    1.422 mixins.py:189(_compute)
      253    0.003    0.000    2.252    0.009 statistical.py:515(compute)
```

**Comment by ssanderson** [comment]
Hrm, though, thinking about this a bit more, I think we either need to require a minimum number of data points if we're handling nans in this way (as @ahgnaw suggests above).  If we don't, we'll produce crazy values for regressions where we only have 2-3 data points.

**Comment by jmccorriston** [comment]
Oh, that's not great. I like Ana's idea of a minimum number or minimum percentage of non-nan date pairs.

**Comment by dmichalowicz** [comment]
@ssanderson Let me know if this needs another round of review

File: zipline/pipeline/factors/statistical.py [hunk 3]
```diff
@@ -475,3 +480,142 @@ def __new__(cls,
...
+
+    def __repr__(self):
+        return "{}(window_length={}, allowed_missing={})".format(
+            type(self).__name__,
+            self.window_length,
+            self.params['allowed_missing'],
+        )
+
+
+def vectorized_beta(dependents, independent, allowed_missing, out=None):
```
Comment (ahgnaw) [inline zipline/pipeline/factors/statistical.py]: I couldn't find another place where `vectorized_beta` might be useful at the moment, however at a later date we might want to move it to `empyrical`.

File: tests/pipeline/test_statistical.py [hunk 4]
```diff
@@ -802,3 +879,139 @@ def test_factor_regression_method_two_factors(self, regression_length):
...
+                expect_nan = num_nans[i] > allowed_missing
+                true_beta = true_betas[i]
+                if expect_nan:
+                    self.assertTrue(np.isnan(result))
+                else:
+                    self.assertTrue(np.abs(result - true_beta) < 0.01)
+
+    def test_allowed_missing_doesnt_double_count(self):
+        # Test that allowed_missing only counts a row as missing one
+        # observation if it's missing in both the dependent and independent
```
Comment (ahgnaw) [inline tests/pipeline/test_statistical.py]: should "both" be "either"? Dates for which the dependent or independent (or both) data is nan cannot be included in the beta calculation.

**Comment by ssanderson** [comment]
@dmichalowicz yeah, I think this is ready for another pass if you can take another look.

File: zipline/pipeline/factors/statistical.py (hunk 3 above)
Comment (ssanderson) [reply zipline/pipeline/factors/statistical.py]: I don't think empyrical would use the `vectorized` part of this anywhere right now, but I suspect that the direct calculation of covariance and variance here is faster than what we currently do in empyrical, because it avoids making copies of the data and it avoids calculating parts of the covariance matrix that we don't need.  I'm not super worried about porting this in the near term, because I don't think anything is currently bottlenecked on empyrical's beta calculations, but if that becomes a bottleneck this is probably a good place to look to speed things up.

File: tests/pipeline/test_statistical.py (hunk 4 above)
Comment (ssanderson) [reply tests/pipeline/test_statistical.py]: I think this wording is correct; the thing that this test is checking is that if a row has a NaN in both the dependent and independent, then we only count that row as adding 1 toward the number of required missing values.

File: zipline/pipeline/factors/statistical.py [hunk 5]
```diff
@@ -475,3 +480,146 @@ def __new__(cls,
...
+    # Cython.
+
+    # shape: (N, M)
+    independent = np.where(
+        isnan(dependents),
+        nan,
+        independent,
+    )
+
+    # Calculate beta as Cov(X, Y) / Cov(Y, Y).
```
Comment (dmichalowicz) [inline zipline/pipeline/factors/statistical.py]: I think you mean Cov(X, Y) / Cov(X, X)  (The calculation below is correct, just this comment is off)

File: zipline/pipeline/factors/statistical.py [hunk 6]
```diff
@@ -475,3 +480,146 @@ def __new__(cls,
...
+    # shape: (M,)
+    covariances = nanmean(ind_residual * dep_residual, axis=0)
+
+    # We end up with different variances in each column here because each
+    # column may have a different subset of the data dropped due to missing
+    # data in the corresponding dependent column.
+    # shape: (M,)
+    independent_variances = nanmean(ind_residual ** 2, axis=0)
+
+    # shape: (M,)
```
Comment (dmichalowicz) [inline zipline/pipeline/factors/statistical.py]: Thanks for these shape comments!

File: tests/pipeline/test_statistical.py [hunk 7]
```diff
@@ -802,3 +879,139 @@ def test_factor_regression_method_two_factors(self, regression_length):
...
+        noise = as_column(rand.uniform(-.1, .1, 30))
+        dependents = 1.0 + true_betas * independent + noise
+
+        # Fill 20% of the input arrays with nans randomly.
+        dependents[rand.uniform(0, 1, dependents.shape) < 0.2] = nan
+        independent[rand.uniform(0, 1, independent.shape) < 0.2] = nan
+
+        # Sanity check that we actually inserted some nans.
+        self.assertTrue(np.count_nonzero(np.isnan(dependents)) > 0)
+        self.assertTrue(np.count_nonzero(np.isnan(independent)) > 0)
```
Comment (dmichalowicz) [inline tests/pipeline/test_statistical.py]: It's a very small chance, but doesn't this mean this test could fail randomly?

File: tests/pipeline/test_statistical.py [hunk 8]
```diff
@@ -802,3 +879,139 @@ def test_factor_regression_method_two_factors(self, regression_length):
...
+                     np.array([False, True, True, True, False]))
+
+        # With four allowed missing values, everything but the middle column
+        # should produce a value. The middle column will have 5 nans because
+        # the dependent nans have no overlap with the independent nans.
+        result5 = vectorized_beta(dependents, independent, allowed_missing=4)
+        assert_equal(np.isnan(result5),
+                     np.array([False, False, True, False, False]))
+
+        # With six allowed missing values, everything should produce a value.
```
Comment (dmichalowicz) [inline tests/pipeline/test_statistical.py]: With _five_ allowed...

File: zipline/pipeline/factors/statistical.py [hunk 9]
```diff
@@ -475,3 +480,146 @@ def __new__(cls,
...
+    #
+    # PERF_TODO: We could probably avoid the space blowup by doing this in
+    # Cython.
+
+    # shape: (N, M)
+    independent = np.where(
+        isnan(dependents),
+        nan,
+        independent,
+    )
```
Comment (dmichalowicz) [inline zipline/pipeline/factors/statistical.py]: Should you do a similar operation on the `dependent` matrix? Otherwise it will be using rows in its residual calculation that the independent variable does not use.

File: tests/pipeline/test_statistical.py (hunk 7 above)
Comment (ssanderson) [reply tests/pipeline/test_statistical.py]: We're seeding the rng, so this is still deterministic.

File: zipline/pipeline/factors/statistical.py (hunk 9 above)
Comment (ssanderson) [reply zipline/pipeline/factors/statistical.py]: Centering dependent turns out not to matter.  See 6bacafab65f8278a3730592b334fdd6dda953dea for a lengthy description.  (We discussed this in person, but putting the link here for posterity.)
